Move from Nullable{T} to Union{T, Null}

nalimilan opened the pull request.

quinnj commented:
Nice! This looks awesome!

quinnj commented on src/nullablearray.jl:
```diff
@@ -1,63 +1,63 @@
 
-@compat (::Type{NullableCategoricalVector{Nullable{T}}}){T}(m::Int; ordered=false) =
+@compat (::Type{NullableCategoricalVector{Union{T, Null}}}){T}(m::Int; ordered=false) =
     NullableCategoricalArray{T}((n,); ordered=ordered)
```
I get a `ERROR: UndefVarError: n not defined` for this method

quinnj replied:
I needed to change this to 
```julia
@compat (::Type{NullableCategoricalVector{Union{T, Null}}}){T}(m::Int; ordered=false) =
    NullableCategoricalArray{T}(m; ordered=ordered)
```

nalimilan replied:
Yeah, good catch. Would you make a PR against master?

quinnj commented:
Using this branch seems to break Base indexing somehow
```julia
julia> using CategoricalArrays

julia> t = Any[1,2,3]
3-element Array{Any,1}:
 1
 2
 3

julia> t[[1,2]]
ERROR: MethodError: Cannot `convert` an object of type Tuple{Base.OneTo{Int64}} to an object of type Array{Any,N} where N
This may have arisen from a call to the constructor Array{Any,N} where N(...),
since type constructors fall back to convert methods.
Stacktrace:
 [1] macro expansion at ./multidimensional.jl:458 [inlined]
 [2] _unsafe_getindex(::IndexLinear, ::Array{Any,1}, ::Array{Int64,1}) at ./multidimensional.jl:453
 [3] macro expansion at ./multidimensional.jl:442 [inlined]
 [4] _getindex at ./multidimensional.jl:438 [inlined]
 [5] getindex(::Array{Any,1}, ::Array{Int64,1}) at ./abstractarray.jl:875

```

quinnj commented on REQUIRE:
```diff
@@ -1,3 +1,3 @@
 julia 0.5.0
```
Probably should bump the julia requirement here to `julia 0.6-`

quinnj commented on src/typedefs.jl:
```diff
@@ -73,7 +75,7 @@ end
 ## Nullable Arrays
 
-@compat abstract type AbstractNullableCategoricalArray{T, N, R} <: AbstractArray{Nullable{CategoricalValue{T, R}}, N} end
+@compat abstract type AbstractNullableCategoricalArray{T, N, R} <: AbstractArray{Union{CategoricalValue{T, R}, Null}, N} end
```
So to remind myself, the idea you mentioned was to actually get rid of `AbstractNullableCategoricalArray`? And just have
```julia
@compat abstract type AbstractCategoricalArray{T, N, R} <: AbstractArray{Union{CategoricalValue{T, R}, Null}, N} end
```
?
I'm currently building out a `WeakRefStringArray` type and trying to get the subtyping right. `getindex` would then have a return type of `Union{WeakRefString{T}, Null}`.

nalimilan replied:
Actually the only solution I could find is to have
```julia
abstract type AbstractCategoricalArray{T, N, R, V, U} <: AbstractArray{Union{CategoricalValue{V, R}, U}, N} end
immutable CategoricalArray{T, N, R <: Integer, V, U} <: AbstractCategoricalArray{T, N, R, V, U} [...] end
```
Then a non-nullable `CategoricalArray` will look like `CategoricalArray{String, 1, UInt32, String, Union{}}`, and `NullableCategoricalArray` will be replaced with `CategoricalArray{Union{String, Null}, 1, UInt32, String, Null}`.

I'm not really happy about the need for the two last parameters, which are redundant with `T`, but they are required (AFAICT) to define the type of the pool and the element type when subtyping `AbstractArray`. At least they are mostly invisible to the user, since the inner constructor computes them automatically and they are not needed for dispatch (we could even avoid printing them).

I can push a mostly working version of this later today if you're interested.

Regarding `WeakRefStringArray`, I'm not sure how it relates to categorical arrays. Do you want it to be a subtype of `AbstractCategoricalArray`?

nalimilan replied:
I've just pushed a new commit which passes most tests and should be OK from an API point of view, but needs more work to clean many things.

nalimilan commented:
@quinnj I've updated the PR a bit and added docs. I think it is now ready: once you can confirm your DataTables branch works with the removal of `NullableCategoricalArray`, I think we can merge this one, and tag a breaking release.

quinnj commented:
Excellent. I have quite a bit of time next week where I can clean up the DataTables PR and ensure things are working w/ this latest branch. I've almost got CSV all the way overhauled w/ new DataStreams API, so I'm a bit all over the place; but all these are good changes and going to lead to much better usability/performance pretty soon.

quinnj commented:
Is this some kind of weird Base printing issue where it auto shows the eltype of an array?
```julia
julia> source.columns
(event = Float64[], mag = Float64[], station = Union{CategoricalArrays.CategoricalValue{String,Int32}, Nulls.Null}[], dist = Float64[], accel = Float64[])

julia> source.columns.station
0-element CategoricalArrays.CategoricalArray{Union{Nulls.Null, String},1,Int32,String,Nulls.Null}
```

nalimilan commented:
You mean, the `Union{CategoricalArrays.CategoricalValue{String,Int32}, Nulls.Null}[]` part? Yeah, that's an old issue, which is quite annoying, especially since that's supposed to be a compact display. I guess we should override the method to just print `(?String)[]`. That's a bit of a lie, but for a compact printing `CategoricalValue` is really too long.

quinnj commented:
Yeah, it was just the auto-display of `eltype(A)` that threw me off at first (I thought I was somehow not actually getting a CategoricalArray allocated). I don't think CategoricalValue is too bad, but it'd be nice if it printed like `CategoricalValue{?String}` instead, the underlying ref type is not really that important.

I think I mentioned it before, but I tried printing `Union{T, Null}` as `?T` at one point in `Nulls.jl`, but ran into some weird segfaulting behavior. Do you want to take a stab at it and see if you have any issues? If not, I'll take another stab

nalimilan commented:
Yes, one solution would be to always print `CategoricalValue{T}` and omit the reference type, which is confusing anyway. Maybe even omit `String`, since non-string categorical values aren't that common.

I can have a look at the `Union{T, Null}` printing issue, but if you have already done some experimentations you must know better than I do. I'll spend my time better improving the printing of `CategoricalValue`.

quinnj commented:
Sounds good; I actually was able to come up with a pretty minimal repro of the segfault and reported it, so we'll see if someone else has a better idea of why `?T` is crashing everything.

nalimilan commented:
It turns out it's not easy to improve the compact printing of `CategoricalArray`. Unless we completely override the `show` method, the element type is printed simply as `string(eltype(X))`. So we would need to change the `show(::IO, ::Type{<:CategoricalValue})`, which doesn't sound like a good idea. We'd better try to find a generic mechanism in Base to allow for a more compact output.

cjprybol commented on docs/src/using.md:
```diff
@@ -65,17 +66,19 @@ The `CategoricalArray` still considers `"Old"` as a possible level even if it is
 
 julia> droplevels!(x)
-2-element Array{String,1}:
+4-element CategoricalArrays.CategoricalArray{String,1,UInt32,String,Union{}}:
```
Is this behavior change intended? Looks like before droplevels! was returning the levels remaining after the update and here it is returning the full CategoricalArray

nalimilan replied:
Yes, I've changed that for consistency, since generally mutating functions return their input. You can easily wrap the call inside `levels(...)` if needed, but the contrary wasn't possible.

cjprybol commented on docs/src/using.md:
```diff
@@ -102,15 +102,17 @@ levels
+The examples above assumed that the data contained no missing values. This is generally not the case in real data. This is where `CategoricalArray{?T}` comes into play. It is essentially the categorical-data equivalent of `Array{?T}`. It behaves exactly the same as `CategoricalArray{T}` , except that when indexed it returns either a `CategoricalValue{T}`, or `null` if a value is missing. See [the ...
 
-Let's adapt the example developed above to support missing values. At first sight, not much changes: 
+Let's adapt the example developed above to support missing values. Since there are no missing values in the input vector, we need to specify that the array should be able to hold eithera `String` or `null`. Here, `?String` is a shorthand for `Union{Null, String}`:
```
eithera -> either a

nalimilan replied:
@quinnj It would be great if you could fix this one for me at the same time as the indentation issues I noted. BTW, I guess `?String` should be removed just like you did in the rest of the code?

quinnj commented:
I rebased this branch and made some 0.7 syntax updates.

ararslan commented on src/nullablearray.jl:
```diff
@@ -1,78 +1,46 @@
-    end
+## (special methods for AbstractArray{Union{T, Null}}, to avoid wrapping nulls inside CategoricalValues)
+
+(::Type{CategoricalArray{Union{T, Null}, N, R}})(dims::NTuple{N,Int}; ordered=false) where {T, N, R} =
```
Since this package is 0.6-only, we can dump the `(::Type{...})(...)` syntax

nalimilan replied:
What's the new syntax? Just inner constructors?

nalimilan commented on src/extras.jl:
```diff
@@ -58,9 +58,9 @@ If `x` is nullable (i.e. `eltype(x) >: Null`), a nullable `CategoricalArray` is
 """
-function cut{T, N, U<:AbstractString}(x::AbstractArray{T, N}, breaks::AbstractVector;
+function cut(x::AbstractArray{T, N}, breaks::AbstractVector;
                                       extend::Bool=false, labels::AbstractVector{U}=String[],
```
Indentation should be adjusted.

nalimilan commented on src/extras.jl:
```diff
@@ -119,6 +119,6 @@ end
-cut{U<:AbstractString}(x::AbstractArray, ngroups::Integer;
-                       labels::AbstractVector{U}=String[]) =
+cut(x::AbstractArray, ngroups::Integer;
+                       labels::AbstractVector{U}=String[]) where {U<:AbstractString} =
```
Same here.

nalimilan commented on src/nullablearray.jl:
```diff
@@ -81,77 +49,27 @@ function NullableCategoricalArray{T, N}(A::AbstractArray{T, N},
-in{T, N, R}(x::Any, y::NullableCategoricalArray{T, N, R}) = false
-in{T, N, R}(x::CategoricalValue, y::NullableCategoricalArray{T, N, R}) = false
+in(x::Null, y::CategoricalArray) = false
+in(x::Null, y::CategoricalArray{<:Null}) = !all(v -> v > 0, y.refs)
```
I've just noticed I made a mistake here, should be `>:Null` AFAICT. Also needs a test.

nalimilan commented:
Any ideas about why tests fail on 0.7? Maybe related to the remaining uses of `?T`?

quinnj commented:
@nalimilan, I pushed a bunch of updates that should address all your comments. All tests are passing except in nullablearrays.jl. There seems to be some kind of `ordered` issue where even `ordered` CategoricalArrays aren't getting sorted correctly. Could you take a look? I may have messed up one of the constructors or something.

rofinn commented:
FWIW, it looks like the default ordering has changed when constructing a `CategoricalArray` and `ordered!` doesn't seem to change that.

```julia
# nl/null
julia> y = CategoricalVector{Union{String, Null}, UInt32}(["b", "a", "b"], ordered=false)
3-element CategoricalArrays.CategoricalArray{?String,1,UInt32,String,Nulls.Null}:
 "b"
 "a"
 "b"

julia> y.pool
CategoricalArrays.CategoricalPool{String,UInt32}(["b","a"])

julia> CategoricalArrays.order(y.pool)
2-element Array{UInt32,1}:
 0x00000001
 0x00000002

julia> y = CategoricalVector{Union{String, Null}, UInt32}(["b", "a", "b"], ordered=false)
3-element CategoricalArrays.CategoricalArray{?String,1,UInt32,String,Nulls.Null}:
 "b"
 "a"
 "b"

julia> CategoricalArrays.order(y.pool)
2-element Array{UInt32,1}:
 0x00000001
 0x00000002

julia> ordered!(y, true)
3-element CategoricalArrays.CategoricalArray{?String,1,UInt32,String,Nulls.Null}:
 "b"
 "a"
 "b"

julia> CategoricalArrays.order(y.pool)
2-element Array{UInt32,1}:
 0x00000001
 0x00000002

julia> y.pool
CategoricalArrays.CategoricalPool{String,UInt32}(["b","a"]) with ordered levels

# v0.1.6julia> y = NullableCategoricalVector{String, UInt32}(["b", "a", "b"], ordered=false)
3-element CategoricalArrays.NullableCategoricalArray{String,1,UInt32}:
 "b"
 "a"
 "b"

julia> y.pool
CategoricalArrays.CategoricalPool{String,UInt32}(["a","b"])

julia> CategoricalArrays.order(y.pool)
2-element Array{UInt32,1}:
 0x00000002
 0x00000001

julia> ordered!(y, true)
3-element CategoricalArrays.NullableCategoricalArray{String,1,UInt32}:
 "b"
 "a"
 "b"

julia> CategoricalArrays.order(y.pool)
2-element Array{UInt32,1}:
 0x00000002
 0x00000001

julia> y.pool
CategoricalArrays.CategoricalPool{String,UInt32}(["a","b"]) with ordered levels
```

cjprybol commented on benchmark/benchmarks.jl:
```diff
@@ -11,9 +12,9 @@ using CategoricalArrays
-    @bench "NullableCategoricalArray" sumequals(nca, Nullable("D"))
+    nca = CategoricalArray(repeat([null; string.('A':'J')], outer=1000))
+    @bench "CategoricalArray{String}" sumequals(ca, "D")
+    @bench "CategoricalArray{?String}" sumequals(nca, "D")
```
{?String} -> Union{String, Null}

cjprybol commented on benchmark/benchmarks.jl:
```diff
@@ -26,7 +27,7 @@ end
-    @bench "NullableCategoricalArray" sumequals(nca, nca[1])
+    nca = CategoricalArray(repeat([null; string.('A':'J')], outer=1000))
+    @bench "CategoricalArray{String}" sumequals(ca, ca[1])
+    @bench "CategoricalArray{?String}" sumequals(nca, nca[1])
```
{?String} -> Union{String, Null}

cjprybol commented on docs/src/index.md:
```diff
@@ -1,20 +1,8 @@
-- `CategoricalArray` returns a `CategoricalValue` object
-
-- `NullableCategoricalArray` returns a `Nullable{CategoricalValue}` object
+The package provides the `CategoricalArray` type designed to hold categorical data (either unordered/nominal or ordered/ordinal) efficiently and conveniently. `CategoricalArray{T}` holds values of type `T`. `CategoricalArray{?T}`, a.k.a. `CategoricalArray{Union{Null, T}}`, either holds values of type `T`, or missing values (represented as `null`, of the [`Null`](https://github.com/JuliaData/Nulls ...
```
remove reference to {?T} syntax here

cjprybol commented on docs/src/using.md:
```diff
@@ -102,15 +102,17 @@ levels
+The examples above assumed that the data contained no missing values. This is generally not the case in real data. This is where `CategoricalArray{Union{T, Null}}` comes into play. It is essentially the categorical-data equivalent of `Array{Union{T, Null}}`. It behaves exactly the same as `CategoricalArray{T}` , except that when indexed it returns either a `CategoricalValue{T}`, or `null` if a va ...
 
-Let's adapt the example developed above to support missing values. At first sight, not much changes: 
+Let's adapt the example developed above to support missing values. Since there are no missing values in the input vector, we need to specify that the array should be able to hold eithera `String` or `null`. Here, `?String` is a shorthand for `Union{Null, String}`:
```
?String

cjprybol commented on docs/src/using.md:
````diff
@@ -102,15 +102,17 @@ levels
+```jldoctest using
+julia> using Nulls
+
+julia> y = CategoricalArray{?String}(["Old", "Young", "Middle", "Young"], ordered=true)
````
{?String}

nalimilan commented:
Hmm. Maybe the `method_exists(isless, (T, T))` call in the `convert` method doesn't work when `T = Union{U, Null}`? The levels are only sorted if that's true. We should probably do `method_exists(isless, Tuple{Nulls.T(T), Nulls.T(T)})`. Funny that it wasn't an issue when I first filed the PR.

(To be clear, whether a categorical array is ordered or not does not change its behavior except for the fact that `<` and `>` can be used to compare values.)

nalimilan commented:
Cool, thanks @rofinn! I've included the fix in the original commits so that it passes the test directly. I've also taken into account issues spotted by @cjprybol.

The remaining failure on Julia 0.7 indeed seems like a bug in Julia, as noted on #76. Merging, I'll tag a release once the upper version bound has been added to DataTables.

ararslan commented:
THE FUTURE IS NOW
